registry/handlers/app.go

```diff
@@ -182,100 +166,43 @@ func (app *App) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-}
 
-func (ssrw *singleStatusResponseWriter) Flush() {
-	if flusher, ok := ssrw.ResponseWriter.(http.Flusher); ok {
-		flusher.Flush()
-	}
-}
-
-// dispatcher returns a handler that constructs a request specific context and
-// handler, using the dispatch factory function.
-func (app *App) dispatcher(dispatch dispatchFunc) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		context := app.context(w, r)
-
-		defer func() {
-			ctxu.GetResponseLogger(context).Infof("response completed")
-		}()
+	if app.nameRequired(r) {
+		repository, err := app.registry.Repository(context, getName(context))
 
-		if err := app.authorized(w, r, context); err != nil {
-			ctxu.GetLogger(context).Errorf("error authorizing context: %v", err)
-			return
-		}
-
-		if app.nameRequired(r) {
-			repository, err := app.registry.Repository(context, getName(context))
-
-			if err != nil {
-				ctxu.GetLogger(context).Errorf("error resolving repository: %v", err)
-
-				switch err := err.(type) {
-				case distribution.ErrRepositoryUnknown:
-					context.Errors.Push(v2.ErrorCodeNameUnknown, err)
-				case distribution.ErrRepositoryNameInvalid:
-					context.Errors.Push(v2.ErrorCodeNameInvalid, err)
-				}
-
-				w.WriteHeader(http.StatusBadRequest)
-				serveJSON(w, context.Errors)
-				return
-			}
+		if err != nil {
+			ctxu.GetLogger(context).Errorf("error resolving repository: %v", err)
 
-			// assign and decorate the authorized repository with an event bridge.
-			context.Repository = notifications.Listen(
-				repository,
-				app.eventBridge(context, r))
-
-			context.Repository, err = applyRepoMiddleware(context.Repository, app.Config.Middleware["repository"])
-			if err != nil {
-				ctxu.GetLogger(context).Errorf("error initializing repository middleware: %v", err)
-				context.Errors.Push(v2.ErrorCodeUnknown, err)
-				w.WriteHeader(http.StatusInternalServerError)
-				serveJSON(w, context.Errors)
-				return
+			switch err := err.(type) {
+			case distribution.ErrRepositoryUnknown:
+				return nil, NewHTTPError(v2.ErrorCodeNameUnknown, err, http.StatusBadRequest)
+			case distribution.ErrRepositoryNameInvalid:
+				return nil, NewHTTPError(v2.ErrorCodeNameInvalid, err, http.StatusBadRequest)
 			}
 		}
 
-		handler := dispatch(context, r)
-
-		ssrw := &singleStatusResponseWriter{ResponseWriter: w}
-		handler.ServeHTTP(ssrw, r)
+		// assign and decorate the authorized repository with an event bridge.
+		context.Repository = notifications.Listen(
+			repository,
+			app.eventBridge(context, r))
 
-		// Automated error response handling here. Handlers may return their
-		// own errors if they need different behavior (such as range errors
-		// for layer upload).
-		if context.Errors.Len() > 0 {
-			if ssrw.status == 0 {
-				w.WriteHeader(http.StatusBadRequest)
-			}
-			serveJSON(w, context.Errors)
+		context.Repository, err = applyRepoMiddleware(context.Repository, app.Config.Middleware["repository"])
```

Comment:
This can possible set `Repository` to `nil` if the `applyRepoMiddelware` is misbehaved. The error should be checked before setting the field.